### PR TITLE
Preprocessing pipeline for all data

### PR DESCRIPTION
**Pull request by Sh-imaa**
- Adding elevation and slope  to the pipeline
- Make the pipeline support multiple years/countries, with the ability to process only a subset or all.
- Ability to train on subset of country/year, with the availability of all the data 
- Add augmentation
- Adding borders to normalization  [https://github.com/Sh-imaa/glacier_mapping/issues/12](url)
- Extending readme

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,4 +92,60 @@ for example. The `-c` tells you the path to the config, and `-o` says where to w
 ## looking at results
 
 The results get saved to the output directory. You can use `wandb sync` to upload the results to some [pretty viewers](https://app.wandb.ai/krisrs1128/glacier_mapping/runs/rsii7qj6?workspace=default).
+
+## pre-processing and slicing
+To the run the preprocessing you need two folders, `img_data` and `vector_data`, which contains the tiffs and the labels respectively. The folders should be organized as illustrated bellow. 
+```
+data\
```
Comment (krisrs1128) [inline README.md]: Just to be consistent with path names elsewhere, can you change these to /?

File: README.md [hunk 2]
```diff
@@ -92,4 +92,60 @@ for example. The `-c` tells you the path to the config, and `-o` says where to w
...
+    2010\
+      nepal\ # nepal labels
+      ....
+    2000\
+      ....
+    basin\
+      data\ #.shp files for test labels
+    borders\
+      nepal\
+        data\ #.shp files to nepal borders
```
Comment (krisrs1128) [inline README.md]: Could you add links to these vector data? Not here, but somewhere in the readme?

File: README.md [hunk 3]
```diff
@@ -92,4 +92,60 @@ for example. The `-c` tells you the path to the config, and `-o` says where to w
...
+
+Example:
+
+`year: ['2010'] country:[]`
+will process all countries in 2010.
+
+You should end up with `slices` directory for all the sliced data and `sat_data.csv` for the metadata needed for training. The folder `sat_files` will be created for all intermediate results, for each set of year/country,  to help with future debugging and ease of incremental adding of data. This folder is of no importance for training.
+
+The filtering of data depends, for now, on the amount of labels existing inside the countries border, amount of values out of the satellite image (nan), and amount of labels in general in the slice. Any slice that returns `False` for any of the conditions in the config file, will be marked as invalid. Development split is generated at random, while test split can be random or according to the intersection with the test vector data (check config file for more details)  
+
```
Comment (krisrs1128) [inline README.md:134]: missing period at end of previous sentence

File: README.md [hunk 4]
```diff
@@ -92,4 +92,60 @@ for example. The `-c` tells you the path to the config, and `-o` says where to w
...
+This is a necessary step to train on pseudo labels, and can be skipped with regular glaciers.
+
+## normalization
+To generate `normalization_data.pkl`, you should run the following command
+`python -m src.normalize --base_dir ./data`
+
+The command expects `slices` and `sat_data.csv` to be previously generated.
+
+The normalization operates on all channels/snow_index/borders/elevation/slope, and the right normalization factors will be picked up correctly relative to your training configuration.
+
```
Comment (krisrs1128) [inline README.md:150]: I like how detailed the documentation here is.

File: src/augmentations.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+def rotate(img, mask, rot=(-10, 10), p=0.5):
+    """Rotate image and crossponding mask with the same random angle"""
+
+    if random.random() > p:
+        angle = random.uniform(rot[0], rot[1])
+        w, h = mask.shape
+        center = int(w / 2), int(h / 2)
+
+        rot_mat = cv2.getRotationMatrix2D(center, angle, 1)
+        rotated_img = cv2.warpAffine(img, rot_mat, (w, h))
```
Comment (krisrs1128) [inline src/augmentations.py]: Wouldn't it be more concise to redefine img and mask during the if, and then always return img and mask outside?

File: src/augmentations.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    img = img.data.numpy()
+    img = np.moveaxis(img, 0, -1)
+
+    return img
+
+def rotate(img, mask, rot=(-10, 10), p=0.5):
+    """Rotate image and crossponding mask with the same random angle"""
+
+    if random.random() > p:
```
Comment (krisrs1128) [inline src/augmentations.py]: Isn't this a sort of unintuitive definition of p? I feel like people would assume it's the rotation probability, but isn't this 1 - the augmentation probability?

File: src/dataset.py [hunk 7]
```diff
@@ -9,24 +9,33 @@
...
+            self.data = self.data[self.data["country"].isin(country)]
+        if year != 'all':
+            self.data = self.data[self.data["year"].isin(year)]
         self.img_transform = img_transform
         self.borders = borders
         self.use_cropped = use_cropped
         self.use_snow_i = use_snow_i
-        self.channels_to_inc = channels_to_inc
+
+        if channels_to_inc is not None: self.channels_to_inc = channels_to_inc[:]
```
Comment (krisrs1128) [inline src/dataset.py]: I feel like it's more conventional to indent here, even though I get that this is more concise.

File: src/dataset.py [hunk 8]
```diff
@@ -44,15 +53,21 @@ def __getitem__(self, i):
...
+            mask_path = cropped_label_path
             image_path = cropped_img_path
 
         img = np.load(image_path)
         img = T.ToTensor()(img)
 
-        if self.channels_to_inc is not None:
-            img = img[self.channels_to_inc]
+        # get snow index before filtering the data
+        snow_i = utils.get_snow_index(img)
```
Comment (krisrs1128) [inline src/dataset.py]: why not just call it `snow_index` from here, so that you don't need to redefine it later?

File: src/dataset.py [hunk 9]
```diff
@@ -61,43 +76,61 @@ def __getitem__(self, i):
...
-            snow_index = utils.get_snow_index(img)
-            snow_index = np.expand_dims(snow_index, axis=0)
-            img = np.concatenate((img, snow_index), axis=0)
-            img = torch.from_numpy(img)
-
         if self.img_transform is not None:
             img = self.img_transform(img)
 
-        # default is 'glaciers' for original labels 
+        # default is 'glaciers' for origina/labels
```
Comment (krisrs1128) [inline src/dataset.py]: I think there was a find and replace issue here or something.

File: src/augmentations.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+def to_numpy_img(img):
+    """From tensor to numpy, with channels last"""
+
+    img = img.data.numpy()
+    img = np.moveaxis(img, 0, -1)
+
+    return img
+
+def rotate(img, mask, rot=(-10, 10), p=0.5):
```
Comment (krisrs1128) [inline src/augmentations.py:13]: Is there a reason not to use the transforms in torchvision?

https://pytorch.org/docs/stable/torchvision/transforms.html

File: src/trainer.py [hunk 11]
```diff
@@ -65,13 +65,14 @@ def train_epoch(self, op, loss_f):
...
     """Evaluate a dataset and return loss and metrics."""
+    
     epoch_loss = 0
     self.model.eval()
 
     data = getattr(self, f"{mode}_data")
     epoch_metrics = defaultdict(int)
     wandb_imgs = []
-
+    n = len(data.dataset)
```
Comment (krisrs1128) [inline src/trainer.py]: do you actually use this?

File: src/slice_n_preprocess.py [hunk 12]
```diff
@@ -28,39 +63,31 @@
...
+                               country, year, data_c, valid_c, split_c)
+            
         sat_path = base_dir / "sat_files"
         dfs = []
         for year_path in sat_path.iterdir():
             for country_path in year_path.iterdir():
-                dfs.append(pd.read_csv(country_path / 'sat_data.csv'))
-        pd.concat(dfs).to_csv(base_dir / 'sat_data.csv')
\ No newline at end of file
+                if (country_path / 'sat_data.csv').exists():
```
Comment (krisrs1128) [inline src/slice_n_preprocess.py:91]: Can't you just use name_of_df.to_csv(..., mode="a") to append? Instead of reading in, concatenating, and then writing?

**Review by krisrs1128 — APPROVED** [review]
A few comments related to conciseness / readability, but overall everything makes sense. Will try rerunning once data is on the cluster.

File: src/augmentations.py (hunk 10 above)
Comment (Sh-imaa) [reply src/augmentations.py:13]: 12 channel input

Comment (krisrs1128) [reply src/augmentations.py:13]: Ah, okay, makes sense.

File: src/augmentations.py (hunk 6 above)
Comment (Sh-imaa) [reply src/augmentations.py]: It's the rotation probability. There's no augmentation probability, every operation has its own p, for no augmentation, one needs to set all to 0.
Can add another parameter to set all to 0, when augmentation=false

File: src/slice_n_preprocess.py (hunk 12 above)
Comment (Sh-imaa) [reply src/slice_n_preprocess.py:91]: That would be great, but it would result in repeated rows in case of running a file more than one. Which happens a lot, in development at least. This is the easiest way not to over complex the code and have some flexibility.

File: src/augmentations.py (hunk 6 above)
Comment (krisrs1128) [reply src/augmentations.py]: Sorry, I meant rotation probability. But what I mean is that, if I set p = 0.9 in the argument to this function, won't this return a rotated result with probability 0.1?

File: src/trainer.py (hunk 11 above)
Comment (Sh-imaa) [reply src/trainer.py]: No. Was left over from before. Thanks!

File: src/dataset.py (hunk 9 above)
Comment (Sh-imaa) [reply src/dataset.py]: Thanks!

File: src/slice_n_preprocess.py (hunk 12 above)
Comment (krisrs1128) [reply src/slice_n_preprocess.py:91]: Ah I see, didn't think about repeated data, this makes sense.

File: src/dataset.py (hunk 8 above)
Comment (Sh-imaa) [reply src/dataset.py]: Just didn't notice. Thanks

File: src/augmentations.py (hunk 5 above)
Comment (Sh-imaa) [reply src/augmentations.py]: True

File: src/augmentations.py (hunk 6 above)
Comment (Sh-imaa) [reply src/augmentations.py]: Got you. You're right, it's toggled, will make it the other way around
